### DISCUSSION TITLE
A question on cpu timing counters analysis

### DISCUSSION THREAD

**Discussion by mmcheraghi** [stnolting/neorv32#923] [Q&A]
Hi @stnolting
 
Is it correct to say that the total execution cycles (**HPMCNT_EVENT_CY**) of a simple code is composed of wait cycles for LSU (**HPMCNT_EVENT_WAIT_LSU**) and wait cycles for ALU (**HPMCNT_EVENT_WAIT_ALU**) ?

I'm looking for the right set of HPM counters to understand the contribution of main operation types (memory, alu, ... ) in total execution cycles. 
I expect to be able to sum up partial execution cycles to get the total execution cycles.

Thanks

**Comment by stnolting** [comment]
Basically, yes. But don't forget the wait cycles for the instruction fetch (`HPMCNT_EVENT_WAIT_DIS`). However, all these wait cycles can also overlap as the CPU has **independent** ports and stages for data access and instruction fetch.

**Reply by mmcheraghi** [reply]
Thanks. I'll try it and discuss the results here.

**Reply by mmcheraghi** [reply]
I tried the three mentioned events and the sum is less than total execution cycles.
Does HPMCNT_EVENT_WAIT_ALU dismiss `I` instructions? How about branch instructions?

**Reply by stnolting** [reply]
> Does HPMCNT_EVENT_WAIT_ALU dismiss I instructions? How about branch instructions?

That is correct. This event only track instruction that need **additional** cycles; e.g. all ALU co-processor instructions like shifts, multiplications or division.

To be honest, it is quite hard to "decode" the value of the HPMCNT_EVENT_CY-triggered counter.

Here is a simple example to illustrate this:

```asm
li    a5, -1
li    a6, 0
csrw  mcountinhibit, a5 // enable counters
--------------          // begin timing
lw   ...                // 4 cycles + 1 cycle memory latency
lui  ...                // 2 cycles
add  ...                // 2 cycles
xor  ...                // 2 cycles
sw   ...                // 4 cycles + 1 cycle memory latency
--------------          // end timing
csrw  mcountinhibit, a3 // stop counters (3 cycles)
```

This gives the following HPM results:

```
HPMs:
#00 clock cycles  : 19
#02 instructions  : 6
#03 compr. instr. : 0
#04 DISP waits    : 0
#05 ALU waits     : 0
#06 branch instr. : 0
#07 ctrl flow tr. : 0
#08 MEM loads     : 1
#09 MEM stores    : 1
#10 MEM waits     : 2
#11 traps         : 0
```

In total, 19 cycles are required to process this piece of code consisting of 6 instructions (disabling the counters adds an offset of 3 cycles). 

The three simple ALU instructions (lui, add, xor) need two cycles to complete. There are no wait cycles as these instructions are not processed by any CPU ALU co-processor.

We have two memory accesses (one load and one store) each having one additional cycle for the memory latency ("MEM waits").

So in total we have 19 (= HPMCNT_EVENT_CY) cycles.

Maybe the HPMs in their current state are too imprecise for a good timing/benchmarking analysis...  Do you have any ideas or suggestions what we could improve here? :thinking:

**Reply by mmcheraghi** [reply]
To be honest I don't have any idea currently. But I think the supported events are sufficient totally.

By the way, you solved my problem. I was comparing two decompression algorithms using the same data with different compression rates. The one with fewer compression rate required fewer instructions and more data loads, and the other one required more extraction instructions and fewer data load instructions. 

I thought the second one (with higher compression rate) should run faster but after watching at an HPM counter with `HPMCNT_EVENT_WAIT_DIS` event, I became aware that both data and instruction memories were **external** and the fewer data accesses were compensated by more instruction fetch accesses!

**Reply by mmcheraghi** [reply]
I think we first need to classify cpu operations, e.g. memory access (instruction/data), arithmetic operations (ALU multi-cycle/single cycle, co-processor), control flow (Branch), ... . 

Then we can decide to define some new events.

**Reply by stnolting** [reply]
We already have a trigger for control flow operations: `HPMCNT_EVENT_LOAD` and `HPMCNT_EVENT_STORE` track the number of executed load/store operations.

The amount of "instruction fetch memory accesses" is equal to the number of executed instructions (`HPMCNT_EVENT_IR`) since every instruction was fetched at some time. :wink:

Maybe it is a nice idea to add further events for counting specific instructions types... :thinking:
